edx/analytics/tasks/course_video.py

```diff
@@ -0,0 +1,667 @@
+    Rounds date down to nearest day,
+    and reports the date in a common
+    dd-mm-yyyy format.
+
+    Returns None on error
+    """
+    if type(datetime_obj) != datetime:
+        return None
+    else:
+        return datetime_obj.replace(hour=0, minute=0, second=0, microsecond=0).strftime(DATE_FMT)
+
+
+class CourseVideoEventMixin(object):
+    """
+    Base class for extracting video play information from event.logs.
+    The most basic implementation counts, for each video in each
+    course, the number of unique users and number of times the
+    video has been played.
+    """
+
+    def mapper(self, line):
+        """ Docstring """
+        parsed_tuple_or_none = self.parse_log_line(line)
+        if parsed_tuple_or_none is not None:
+            yield parsed_tuple_or_none
+
+    def reducer(self, key, values):
+        """
+        Base classes should implement this method
+        """
+        raise NotImplementedError
+
+    def parse_log_line(self, line):
+        """ Docstring """
+        try:
+            event = json.loads(line)
+        except ValueError:
+            log.error("Error parsing log line %s" % line)
+            return None
+
+        key = self.get_mapper_key(event)
+        value = self.get_mapper_value(event)
+        if key is None or value is None:
+            return None
+        return key, value
+
+    def get_mapper_key(self, event):
+        """
+        Outputs the key used by the mapper and reducer.
+        Base classes should implement this method.
+        """
+        raise NotImplementedError
+
+    def get_mapper_value(self, event):
+        """
+        Outputs the other tuple used by the mapper.
+        Reducer is sent an iterable of these tuples.
+        Base classes should implement this method.
+        """
+        raise NotImplementedError
+
+
+class CourseVideoSummaryMixin(CourseVideoEventMixin):
+    """ Docstring """
+
+    def reducer(self, key, value_gen):
+        """ Docstring """
+        values = [x for x in value_gen]
+        usernames = map(itemgetter(0), values)
+        unique_users = len(set(usernames))
+        total_activity = len(values)
+        output = (unique_users, total_activity)
+        yield key, output
+
+    def get_mapper_key(self, event):
+        """ Docstring """
+        if type(event) != dict:
+            return None
+
+        video_data = eventlog.get_event_data(event)
+        if video_data is None:
+            return None
+
+        context = event.get('context')
+        if type(context) != dict:
+            return None
+
+        course_id = context.get('course_id')
+        if course_id is None or not opaque_key_util.is_valid_course_id(course_id):
+            return None
+
+        video_id = video_data.get('id')
+        if video_id is None:
+            return None
+
+        timestamp = eventlog.get_event_time(event)
+        if timestamp is None:
+            return None
+
+        date = round_datetime_to_nearest_day(timestamp)
```

Comment:
What is the purpose of rounding to the nearest day? Why not put the actual video access time in the db and then trunc that value to the nearest day if you need to?